datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs

```diff
@@ -615,34 +621,66 @@ mod tests {
         assert_eq!(expect, result);
         // divide
         let left_decimal_array = create_decimal_array(
-            &[Some(1234567), None, Some(1234567), Some(1234567)],
+            &[
+                Some(1234567),
+                None,
+                Some(1234567),
+                Some(1234567),
+                Some(1234567),
+            ],
+            25,
+            3,
+        );
+        let right_decimal_array = create_decimal_array(
+            &[Some(10), Some(100), Some(55), Some(-123), None],
             25,
             3,
         );
-        let right_decimal_array =
-            create_decimal_array(&[Some(10), Some(100), Some(55), Some(-123)], 25, 3);
         let result = divide_decimal(&left_decimal_array, &right_decimal_array)?;
         let expect = create_decimal_array(
-            &[Some(123456700), None, Some(22446672), Some(-10037130)],
+            &[Some(123456700), None, Some(22446672), Some(-10037130), None],
             25,
             3,
         );
         assert_eq!(expect, result);
         let result = divide_decimal_scalar(&left_decimal_array, 10)?;
         let expect = create_decimal_array(
-            &[Some(123456700), None, Some(123456700), Some(123456700)],
+            &[
+                Some(123456700),
+                None,
+                Some(123456700),
+                Some(123456700),
+                Some(123456700),
+            ],
             25,
             3,
         );
         assert_eq!(expect, result);
         // modulus
         let result = modulus_decimal(&left_decimal_array, &right_decimal_array)?;
-        let expect = create_decimal_array(&[Some(7), None, Some(37), Some(16)], 25, 3);
+        let expect =
+            create_decimal_array(&[Some(7), None, Some(37), Some(16), None], 25, 3);
         assert_eq!(expect, result);
         let result = modulus_decimal_scalar(&left_decimal_array, 10)?;
-        let expect = create_decimal_array(&[Some(7), None, Some(7), Some(7)], 25, 3);
+        let expect =
+            create_decimal_array(&[Some(7), None, Some(7), Some(7), Some(7)], 25, 3);
         assert_eq!(expect, result);
 
         Ok(())
     }
+
+    #[test]
+    fn arithmetic_decimal_divide_by_zero() {
+        let left_decimal_array = create_decimal_array(&[Some(101)], 10, 1);
+        let right_decimal_array = create_decimal_array(&[Some(0)], 10, 1);
+
+        let _result = divide_decimal(&left_decimal_array, &right_decimal_array)
+            .expect_err("expecting DivideByZero error");
+        let _result = divide_decimal_scalar(&left_decimal_array, 0)
+            .expect_err("expecting DivideByZero error");
+        let _result = modulus_decimal(&left_decimal_array, &right_decimal_array)
+            .expect_err("expecting DivideByZero error");
+        let _result = modulus_decimal_scalar(&left_decimal_array, 0)
+            .expect_err("expecting DivideByZero error");
```

Comment:
I think these tests would pass even on master without your code change


Perhaps we could add a check for the actual error ? With code similar to https://github.com/apache/arrow-datafusion/blob/331922036a23265084ab091fa2ab5c57da42c841/datafusion/core/tests/sql/information_schema.rs#L35-L42 ?

Reply:
Hah, excellent! Thanks for the review, as always. Have updated with these changes.